webapp/tests/karma/ts/services/enketo.service.spec.ts

```diff
@@ -1124,6 +1124,583 @@ describe('Enketo service', () => {
+          'fields.secret_code_name': 'S4L',
+          'fields.repeat_section[0].extra': 'data1',
+          'fields.repeat_section[0].some.deep.structure.repeat_doc_ref': actual[1]._id,
+          'fields.repeat_section[1].extra': 'data2',
+          'fields.repeat_section[1].some.deep.structure.repeat_doc_ref': actual[2]._id,
+          'fields.repeat_section[2].extra': 'data3',
+          'fields.repeat_section[2].some.deep.structure.repeat_doc_ref': actual[3]._id,
+        });
+      });
+    });
+
+    it('db-doc-ref with repeats with refs outside of repeat', () => {
+      form.validate.resolves(true);
+      const content =
+        `<data xmlns:jr="http://openrosa.org/javarosa">
+            <name>Sally</name>
+            <lmp>10</lmp>
+            <secret_code_name tag="hidden">S4L</secret_code_name>
+            <separate_doc db-doc="true">
+                <type>separat5e</type>
+                <some_property>some_value_1</some_property>
+                <my_parent db-doc-ref="/data"/>
+            </separate_doc>
+              
+            <repeat_section>
+              <extra>data1</extra>             
+              <repeat_doc_ref db-doc-ref="data/separate_doc"/>             
+            </repeat_section>
+            <repeat_section>
+              <extra>data2</extra>             
+              <repeat_doc_ref db-doc-ref="data/separate_doc"/>             
+            </repeat_section>
+            <repeat_section>
+              <extra>data3</extra>             
+              <repeat_doc_ref db-doc-ref="data/separate_doc"/>             
+            </repeat_section>
+          </data>`;
+      form.getDataStr.returns(content);
+
+      dbBulkDocs.resolves([
+        { ok: true, id: '6', rev: '1-abc' },
+        { ok: true, id: '7', rev: '1-def' },
+      ]);
+      dbGetAttachment.resolves(`<form/>`);
+      FileReader.utf8.resolves(`
+        <data>
+          <repeat nodeset="/data/repeat_section"></repeat>
+        </data>
+      `);
+      UserContact.resolves({ _id: '123', phone: '555' });
+      return service.save('V', form).then(actual => {
+        expect(form.validate.callCount).to.equal(1);
+        expect(form.getDataStr.callCount).to.equal(1);
+        expect(dbBulkDocs.callCount).to.equal(1);
+        expect(UserContact.callCount).to.equal(1);
+
+        expect(actual.length).to.equal(2);
+        const doc = actual[0];
+
+        expect(doc).to.deep.nested.include({
+          form: 'V',
+          'fields.name': 'Sally',
+          'fields.lmp': '10',
+          'fields.secret_code_name': 'S4L',
+          'fields.repeat_section[0].extra': 'data1',
+          'fields.repeat_section[0].repeat_doc_ref': actual[1]._id,
+          'fields.repeat_section[1].extra': 'data2',
+          'fields.repeat_section[1].repeat_doc_ref': actual[1]._id,
+          'fields.repeat_section[2].extra': 'data3',
+          'fields.repeat_section[2].repeat_doc_ref': actual[1]._id,
+        });
+      });
+    });
+
```

Comment:
Thanks for the tests. It's much harder to read them when the XML is inlined in a constant, and also it means we can't reuse XML between tests where appropriate. I think this is a good chance to investigate a better way of doing it.

My first thought is to have a factory which returns the string, which has the benefit of being able to potentially pass in variables to modify the XML slightly without having to duplicate it. This is akin to how Rosie works, but for XML files instead of JS objects.

The other idea is to have a folder of files which are read in during the test. This may be slightly slower but I don't think it would be noticeable. The main benefit of this is you get syntax highlighting and errors inline when editing the XML.

The optimal solution may be both - have a factory which reads the files. Either way, let's explore what the ideal solution would look and set up the framework as part of this change, and then future changes can utilise it!

Reply:
I'm not a fan of factories or loaded resources in tests in general. I find it most clear and useful when I can tell what a test is testing with just one glimpse, and without needing to load additional files or resources or guess to what the test-case looks like.

Reply:
I pushed a change to read xml data from xml files. And then builds started failing :)
It turned out to be unrelated, though.